tests/test_interop_scenarios.rs

```diff
@@ -0,0 +1,263 @@
+// The following tests are test corresponding the the interop test scenarios
+// detailed here:
+// https://github.com/mlswg/mls-implementations/blob/master/test-scenarios.md
+// The tests are conducted for every available ciphersuite, but currently only
+// using BasicCredentials. We can change the test setup once #134 is fixed.
+
+fn default_managed_group_config() -> ManagedGroupConfig {
+    let handshake_message_format = HandshakeMessageFormat::Plaintext;
+    let update_policy = UpdatePolicy::default();
+    let callbacks = default_callbacks::default_callbacks();
+    ManagedGroupConfig::new(handshake_message_format, update_policy, 10, callbacks)
+}
+
+/// # 1:1 join
+/// A:    Create group
+/// B->A: KeyPackage
+/// A->B: Welcome
+/// ***:  Verify group state
+ctest!(one_to_one_join {
+    let ciphersuite_name = CiphersuiteName::try_from(_ciphersuite_code).unwrap();
+    println!("Testing ciphersuite {:?}", ciphersuite_name);
+    let ciphersuite = Config::ciphersuite(ciphersuite_name).unwrap();
+    let number_of_clients = 2;
+    let setup = ManagedTestSetup::new(default_managed_group_config(), number_of_clients);
+    setup.create_clients();
+
+    // Create a group with a random creator.
+    let group_id = setup
+        .create_group(ciphersuite)
+        .expect("Error while trying to create group.");
+    let mut groups = setup.groups.borrow_mut();
+    let group = groups.get_mut(&group_id).unwrap();
+
+    let (_, alice_id) = group.members.first().unwrap().clone();
+
+    // A vector including bob's id.
+    let bob_id = setup.random_new_members_for_group(group, 1).unwrap();
+
+    setup
+        .add_clients(ActionType::Commit, group, &alice_id, bob_id)
+        .expect("Error adding Bob");
+
+    // Check that group members agree on a group state.
+    setup.check_group_states(group);
+});
+
+/// # 3-party join
+/// A: Create group
+/// B->A: KeyPackage
+/// A->B: Welcome
+/// C->A: KeyPackage
+/// A->B: Add(C), Commit
+/// A->C: Welcome
+/// ***:  Verify group state
+ctest!(three_party_join {
+    let ciphersuite_name = CiphersuiteName::try_from(_ciphersuite_code).unwrap();
+    println!("Testing ciphersuite {:?}", ciphersuite_name);
+    let ciphersuite = Config::ciphersuite(ciphersuite_name).unwrap();
+
+    let number_of_clients = 3;
+    let setup = ManagedTestSetup::new(default_managed_group_config(), number_of_clients);
+    setup.create_clients();
+
+    // Create a group with a random creator.
+    let group_id = setup
+        .create_group(ciphersuite)
+        .expect("Error while trying to create group.");
+    let mut groups = setup.groups.borrow_mut();
+    let group = groups.get_mut(&group_id).unwrap();
+
+    let (_, alice_id) = group.members.first().unwrap().clone();
+
+    // A vector including Bob's id.
+    let bob_id = setup.random_new_members_for_group(group, 1).unwrap();
+
+    // Create the add commit and deliver the welcome.
+    setup
+        .add_clients(ActionType::Commit, group, &alice_id, bob_id)
```

Comment:
So the way I understand the test case there shouldn't be a commit here but only after everyone is added. Not sure what exactly the difference then is to the next test case though. Either the test or the description need to be updated. I guess the Welcome is supposed to also imply commit.

Reply:
You only get a `Welcome` if you make a commit, so I took `A->B: Welcome` as instruction to actually do an `Add` + `Commit`. But I agree, the instructions could be more precise.